Don't unnecessarily delete folder customization

sledgehammer999 opened the pull request:
Closes #11575

Chocobo1 commented on src/base/utils/fs.cpp:
```diff
@@ -99,9 +99,13 @@ QString Utils::Fs::folderName(const QString &filePath)
+ * `.DS_Store` and/or only temp files that end with '~', e.g. `filename~`.
+ * If they are the only files it will try to remove them and delete the folder.
+ * This action will be performed for each subfolder starting from the deepest folder.
+ * There is an inherent race condition here. A file might appear after it is checked
```
Is the race condition worth mentioning? Nearly all FS operation has race condition.

glassez commented:
Torrent manages files, folders are created as "side effect". So when torrent deletes its content it shouldn't delete any folder if there are 3rd party files inside (except "temporary" files), right? I'm sure, Yes. But in practice different kind of users expect different behavior in this case.

sledgehammer999 commented:
Guys the purpose of smartRemoveEmptyFolderTree() was to remove the folder tree ONLY if it is empty. If it isn't the operation fails by design. The bug here is that we remove the useless files even when the folder isn't going to be deleted (because it isn't empty). It turns out that those files aren't always "useless". If the folder isn't going to be removed then those files must remain too.

sledgehammer999 commented:
`Thumbs.db` -> holds thumbnails
`desktop.ini` -> holds folder customization
`filename.txt~` -> might not be a stale backup. There might currently be a `filename.txt` which is actively edited in an editor. If we remove the backup we fuck up the user.

All the above files (I didn't even check macos) are useful to the user if the folder isn't going to be removed due to 3rd party files. We shouldn't fuck up his folders. Why is it so hard to grasp?

`smartRemoveEmptyFolderTree()` is called after libtorrent deletes the torrent files. It isn't its job to remove torrent files. It is its job to remove empty folders. Here we define empty folders the ones that are either truly empty OR the ones that contain **only** the above mentioned "useless" files.

Chocobo1 commented:
>Here we define empty folders the ones that are either truly empty OR the ones that contain only the above mentioned "useless" files.

And how would this solve the related issue? What if the folder is really empty (except desktop.ini) on the user machine?

So far I'm inclined to think the problem isn't in `smartRemoveEmptyFolderTree()`. It is in the function that calls `smartRemoveEmptyFolderTree()` or upper level's problem.

glassez commented:
>The bug here is that we remove the useless files even when the folder isn't going to be deleted (because it isn't empty). It turns out that those files aren't always "useless". If the folder isn't going to be removed then those files must remain too.

Ok, I understand it now.

glassez commented:
>So far I'm inclined to think the problem isn't in smartRemoveEmptyFolderTree().

Why? `smartRemoveEmptyFolderTree()` shouldn't delete anything in folder if it isn't empty (contains some files except "useless" ones).

Chocobo1 commented:
>Why? 

I think the more important question (for me) is in the first sentence in my post above: https://github.com/qbittorrent/qBittorrent/pull/11660#issuecomment-565708615
The user doesn't state anything about the condition of the affect folder.

Chocobo1 commented:
I'll try to formulate the user's problem: The user has download folder at /a and in it has desktop.ini. And he download a torrent to /a and it resides in /a/b. Then he deletes the torrent and the files, /a/b should be removed and /a/desktop.ini will not be affected.
If the above is correct and is what qbt currently do, then why the user filed a complaint? It must be somewhere else qbt is doing wrong.
Please tell me if I overlooked something.

sledgehammer999 commented:
>And how would this solve the related issue?

Because the user didn't compain that we removed the folder. He complained that we removed his customization. This indicates that the folder stayed because it wasn't empty. (he probably saves other torrents there too).

>Another related question is since /a/b is going to be removed, why would the user care about the desktop.ini in it?

You probably don't understand old and new algorithm. If `b` is empty then `desktop.ini` will be removed. In the new case, the only way `desktop.ini` won't be removed is if there are 3rd party files (unrelated to current torrent and unrelated to the "useless" files).

>The user has download folder at /a and in it has desktop.ini. And he download a torrent to /a and it resides in /a/b. Then he deletes the torrent and the files, /a/b should be removed and /a/desktop.ini will not be affected.


>Then he deletes the torrent and the files, /a/b should be removed and /a/desktop.ini will not be affected.

In both old code `/a/desktop.ini` is removed always. In new code `/a/desktop.ini` will be removed too if there aren't any other user files in folder `a`. In old code if there are other user files in `a`, then `a` will not be removed but the user will lose his customization of `a`.

Chocobo1 commented:
>You probably don't understand old and new algorithm.

I don't think that is the case. In short, you used 2 loops, the first one checks if there is any other files and proceed to delete the folder if there is none (other files).

>In both old code /a/desktop.ini is removed always. 

That shouldn't be the case, as said before, this is a logical error lies elsewhere in qbt and not `smartRemoveEmptyFolderTree()` fault.
For the /a/b case, it should call `smartRemoveEmptyFolderTree("/a/b")`. And for the /a/c case, qbt should not call `smartRemoveEmptyFolderTree()` at all, isn't that correct?
So far IMO this PR is not really addressing the root cause, it is working around the problem (and also adding unnecessary checks to `smartRemoveEmptyFolderTree()`).

sledgehammer999 commented:
I pushed a fix. Now `a/b/desktop.ini` will not be deleted in the following case:
```
a-->b-->c-->desktop.ini
    |   |
    |   |-->3rdparty-file.txt
    |
    |--->desktop.ini

```

sledgehammer999 commented:
>For the /a/b (multi-files) case, the real correct behavior is qbt should call smartRemoveEmptyFolderTree("/a/b"). And for the /a/c (single-file) case, qbt should not call smartRemoveEmptyFolderTree() at all, isn't that correct?

That's not the job of smartRemoveEmptyFolderTree(). (I think you agree too). That's the job of the calling code. And from a very quick glance I think calling code setups correctly in this case too.
And because of this, IMO, it is out of scope for this PR and I don't wish to investigate if calling code does its job correctly.

Chocobo1 commented:
>>For the /a/b (multi-files) case, the real correct behavior is qbt should call smartRemoveEmptyFolderTree("/a/b"). And for the /a/c (single-file) case, qbt should not call smartRemoveEmptyFolderTree() at all, isn't that correct?
>
>And from a very quick glance I think calling code setups correctly in this case too.

If everything works as expected then why the user filed an issue? Ideally /a/dekstop.ini won't ever be deleted. Something is not right clearly.

>That's not the job of smartRemoveEmptyFolderTree(). And because of this, IMO, it is out of scope for this PR and I don't wish to investigate if calling code does its job correctly.

This is exactly what I mean for this PR, as said before I think PR is not addressing that issue directly by fixing the root cause.

Also you should utilize `std::any()` or something similar, other than that there is nothing more for me to say.

sledgehammer999 commented:
>Ideally /a/dekstop.ini won't ever be deleted since smartRemoveEmptyFolderTree() will ever operate on /a

If smartRemoveEmptyFolderTree() operates on `a` then `/a/desktop.ini` will **always** be deleted with old code. Read the old function as if `smartRemoveEmptyFolderTree("a") was called. (read it whole, it doesn't show up all in the diff). But what happens if user has another torrent saved in `a`? Or he himself has pasted another file in `a`? In this 2 cases those files will not be removed but `desktop.ini` will be removed. In new code `desktop.ini` will stay too.

sledgehammer999 commented:
>Also you should utilize std::any() or something similar, other than that there is nothing more for me to say.

Done. Implemeted using `std::all_of()`

Chocobo1 commented:
>If smartRemoveEmptyFolderTree() operates on a then /a/desktop.ini will always be deleted with old code.

My point is `smartRemoveEmptyFolderTree()` should **never** operate on /a in *all* possible cases listed above. I have raised this point many times already, why do you keep saying operating on /a is correct?

>Read the old function as if smartRemoveEmptyFolderTree("a") was called. (read it whole, it doesn't show up all in the diff). 

I heavily refactored/rewritten that function in the past, I know what it is/was doing.

>But what happens if user has another torrent saved in a? 

You must treat each torrent cases individually. Also it depends on whether it is a multi-file torrent or single-file torrent because there are different removal expectations for it.

Chocobo1 commented on src/base/utils/fs.cpp:
```diff
@@ -128,17 +132,26 @@ bool Utils::Fs::smartRemoveEmptyFolderTree(const QString &path)
+            return (deleteFilesList.contains(f, Qt::CaseInsensitive)
+                    || f.endsWith('~'));
+        };
+        if (!std::all_of(tmpFileList.cbegin(), tmpFileList.cend(), predicate))
```
I would store it to a const bool variable instead. The predicate lambda itself doesn't mean much but the result of std::all_of does.

sledgehammer999 replied:
The return value is used only in that if statement. Does it matter if it is stored or not?
Unless, you mean that it improves readability, which I don't think so.

Chocobo1 replied:
It improves readability (for me) and debuggability (that you can type in variable name to watch).

Chocobo1 replied:
```c++
const bool hasOtherFiles = !std::all_of(tmpFileList.cbegin(), tmpFileList.cend(), []()
{
    // ...
});
```

sledgehammer999 replied:
Done.

sledgehammer999 commented:
>should never operate on /a in all possible cases listed above

Can you repeat the cases? Also what does `/a` represent here? Is it part of the save path or part of the torrent file structure (eg root folder)?

Chocobo1 commented on src/base/utils/fs.cpp:
```diff
@@ -118,6 +122,14 @@ bool Utils::Fs::smartRemoveEmptyFolderTree(const QString &path)
+    // temp files on linux usually end with '~', e.g. `filename~`
+    const auto predicate = [&deleteFilesList](const QString &f)
+    {
+        return (deleteFilesList.contains(f, Qt::CaseInsensitive)
```
now this is slow, deleteFilesList should use a QSet instead.

sledgehammer999 replied:
Are you sure? According to [this](https://doc.qt.io/qt-5/containers.html) the lookup for QList is O(1) while the lookup for QSet is amortized O(1). So QList should be faster.

Chocobo1 replied:
>Are you sure? 

Yes I'm sure, I won't bother you with the details as there are many tutorials about algorithm complexity on the net. I will only point out that what you saw O(1) for QList is called *Index* lookup, that is not the same as searching within the container.

Chocobo1 replied:
Also you might want to wrap the entries in deleteFilesList with `QLatin1String()`

Chocobo1 commented:
>Can you repeat the cases? 

https://github.com/qbittorrent/qBittorrent/pull/11660#issuecomment-565711368

>Also what does /a represent here? Is it part of the save path or part of the torrent file structure (eg root folder)?

You can think /a like c:\downloads that doesn't belong to any torrent and only belongs to the user. User can download any torrent structure to /a.

sledgehammer999 commented:
@Chocobo1 I now understand your objections. Here are my findings.
Let's assume that `C:\a` is the save path. Let's also assume these 4 torrent variations:
1. Torrent with multiple files under root folder `b`
2. Torrent with multiple files without creating the root folder (aka files are saved under `C:\a`)
2. Torrent with single file `filename.txt`
3. Torrent with  single file under a root folder `b\filename.txt`

For each of those this is the path that is passed to smart remove:
1. `C:\a\b`
2. `<empty_string>`
3. `C:\a\filename.txt`
4. `C:\a\b`

So in all cases smart remove is called in such a way that nothing inside `C:\a` is removed. I don't know how the user in #11575 has come with a way that `deskop.ini` is removed from his `C:\a`.
So maybe this PR doesn't fix #11575 and I should removed it from the commit message.

**But** this PR is legitimate IMO. It still fixes other corner cases. Aka the case where other files are put there that don't belong to the torrent being removed.

Chocobo1 commented:
>So maybe this PR doesn't fix #11575 and I should removed it from the commit message.

You should definitely do that and revise the commit message to what it really does.

>But this PR is legitimate IMO. It still fixes other corner cases. Aka the case where other files are put there that don't belong to the torrent being removed.

I don't object this PR if given the correct reason (which I learned that the new algorithm is probably the expected behavior from everyone).

Chocobo1 commented on src/base/utils/fs.cpp:
```diff
@@ -128,17 +132,26 @@ bool Utils::Fs::smartRemoveEmptyFolderTree(const QString &path)
-        // remove temp files on linux (file ends with '~'), e.g. `filename~`
         const QDir dir(p);
+        if (!dir.isEmpty(QDir::AllDirs | QDir::NoDotAndDotDot))
+            continue;
```
I would add empty line below, please.

glassez replied:
Sorry, I don't understand this one. Don't you remove folder if it contains subfolders?

sledgehammer999 replied:
Yes. The search is done from deepest folder outwards. But if a deeper folder wasn't removed, then this one shouldn't be removed (or have useless files removed).
```const QStringList tmpFileList = dir.entryList(QDir::Files);```
holds only filenames.

glassez replied:
It would be nice to have such a comment at this point to avoid misunderstandings during future reviews.

sledgehammer999 replied:
Done.

Chocobo1 commented on src/base/utils/fs.cpp:
```diff
@@ -128,17 +132,26 @@ bool Utils::Fs::smartRemoveEmptyFolderTree(const QString &path)
+
+        // deleteFilesList contains unwanted files, usually created by the OS
+        // temp files on linux usually end with '~', e.g. `filename~`
+        const bool hasOtherFiles = !std::all_of(tmpFileList.cbegin(), tmpFileList.cend(), [&deleteFilesList](const QString &f)
```
Pardon me, I still think it can be improved a little bit:
```c++
// use `std::any_of()` instead of `!std::all_of()`
const bool hasOtherFiles = std::any_of(tmpFileList.cbegin(), tmpFileList.cend(), [&deleteFilesList](const QString &f)
{
  // cheaper check goes first
  return (!f.endsWith('~') && !deleteFilesList.contains(f, Qt::CaseInsensitive));
});
```

sledgehammer999 replied:
Ok, I'll do the changes a bit later. But I would suppose that all_of() would have this optimization inside it. (aka break early on the 1st false).

sledgehammer999 commented:
@Chocobo1 I implemented changes apart from one. QStringList::contains() provides a convenience of comparing string without case sensitivity. IMO, this is helpful here. Plus, the container has only 4 elements. The speed gain should be minimal, if any, right?

Chocobo1 commented:
>QStringList::contains() provides a convenience of comparing string without case sensitivity. IMO, this is helpful here. 

Yes you're right on this one and it seems we have no other convenient choice, so just leave it.

Also it seems you still haven't revised the commit message. You could use "Revise smartRemoveEmptyFolderTree()" as title and briefly state why this new behavior is expected.

sledgehammer999 commented:
IMO, the commit title is correct that's why I didn't change it. (Apart from remove the `closes` keyword`).

I know that "customization" doesn't describe all "useless" files. But "Revise ..." is way too generic and probably violates the [guidelines](https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md#10-git-commit-message).

Chocobo1 commented:
@glassez 
This PR is almost ready if you are OK with the commit message, I can approve it after you.

glassez commented on src/base/utils/fs.cpp:
```diff
@@ -128,18 +132,24 @@ bool Utils::Fs::smartRemoveEmptyFolderTree(const QString &path)
+            continue;
+
+        for (const QString &f : tmpFileList)
+            if (f.endsWith('~') || deleteFilesList.contains(f, Qt::CaseInsensitive))
```
Why this additional check? hasOtherFiles is false here so there's no "other" files that shouldn't be deleted, isn't it?

sledgehammer999 replied:
Sorry. This was lost during the many refactorings. Fixed now.

glassez commented:
> if you are OK with the commit message

I would prefer to see something more appropriate. E.g. "Don't remove temporary files if folder will not removed" or "Don't clean up the folder if it isn't really empty".

sledgehammer999 commented:
I changed the commit title. I hope it is ok now.

Chocobo1 commented on src/base/utils/fs.cpp:
```diff
@@ -128,18 +132,25 @@ bool Utils::Fs::smartRemoveEmptyFolderTree(const QString &path)
         const QDir dir(p);
+        // A deeper folder may have not been removed in the previous iteration
+        // so don't remove anything from this folder either.
+        if (!dir.isEmpty(QDir::AllDirs | QDir::NoDotAndDotDot))
```
I overlooked this, wouldn't this be slow since it is going to enumerate all the subfolders every time and count them?
https://doc.qt.io/qt-5/qdir.html#isEmpty

This should overcome that problem:
```c++
bool Utils::Fs::smartRemoveEmptyFolderTree(const QString &path)
{
    /* same code, comment them out
    if (path.isEmpty() || !QDir(path).exists())
        return true;

    const QStringList deleteFilesList = {
        // Windows
        "Thumbs.db",
        "desktop.ini",
        // Linux
        ".directory",
        // Mac OS
        ".DS_Store"
    };

    // travel from the deepest folder and remove anything unwanted on the way out.
    QStringList dirList(path + '/');  // get all sub directories paths
    QDirIterator iter(path, (QDir::AllDirs | QDir::NoDotAndDotDot), QDirIterator::Subdirectories);
    while (iter.hasNext())
        dirList << iter.next() + '/';
    // sort descending by directory depth
    std::sort(dirList.begin(), dirList.end()
              , [](const QString &l, const QString &r) { return l.count('/') > r.count('/'); });
   */

    const bool foundOtherFiles = std::any_of(dirList.cbegin(), dirList.cend(), [&deleteFilesList](const QString &path) {
        const QDir dir(path);
        const QStringList tmpFileList = dir.entryList(QDir::Files);
        return std::any_of(tmpFileList.cbegin(), tmpFileList.cend(), [&deleteFilesList](const QString &f)
        {
            return (!f.endsWith('~') && !deleteFilesList.contains(f, Qt::CaseInsensitive));
        });
    });
    if (foundOtherFiles)
        return false;

    removeDirRecursive(path);
    return QDir(path).exists();
}
```

sledgehammer999 replied:
Aren't you a bit obsessed on not introducing slowness, even in seldomly called functions?
In any case the docs state:
>Equivalent to count() == 0 with filters QDir::AllEntries | QDir::NoDotAndDotDot, **but faster as it just checks whether the directory contains at least one entry.**

I am worried if I should have used `QDir::Dirs` instead. I don't understand the difference between that and `QDir::AllDirs` from the docs.

PS: I can't clearly follow your proposed approached. I don't know if it can account for a branch in the subfolders where one branch isn't empty and the other is. Will it remove the empty branch up to the point where the 2 branches have the common parent-folder?

Chocobo1 replied:
or disregard my comment/code above (not good as now I rethink) and use `if (!dir.isEmpty(QDir::Dirs | QDir::NoDotAndDotDot)) return false;`

sledgehammer999 replied:
You mean:
```c++
 if (!dir.isEmpty(QDir::Dirs | QDir::NoDotAndDotDot))
    continue;
```
right?

Chocobo1 replied:
>I am worried if I should have used QDir::Dirs instead. I don't understand the difference between that and QDir::AllDirs from the docs.

Yes, it seems `QDir::AllDirs` has special effects when using `QDir::entryList()`. I think most other time you should use `QDir::Dirs`.

>Aren't you a bit obsessed on not introducing slowness, even in seldomly called functions?

Sorry, but from algorithm viewpoint the current code feels uneasy for me. A naive example is querying over a network drive with caching disabled, then practically you'll want to avoid any redundant operation.
BTW, IMO performance is always a consideration when writing code and programmer should not easily neglect/trade off performance for nothing.
Taking quote from https://ubiquity.acm.org/article.cfm?id=1513451:
>But, conversely, when designing software at a system level, performance issues should always be considered from the beginning. A good software developer will do this automatically, having developed a feel for where performance issues will cause problems. An inexperienced developer will not bother, misguidedly believing that a bit of fine tuning at a later stage will fix any problems.

Chocobo1 replied:
> You mean:
> 
> ```c++
>  if (!dir.isEmpty(QDir::Dirs | QDir::NoDotAndDotDot))
>     continue;
> ```
> 
> right?

No. Can't you use `return false;` here?

sledgehammer999 replied:
>Not exactly. Can't you use return false; here?

We're inside a loop of folders. Consider this folder tree
>a->b->ca->d->3rd_party_file
>a->b->cb

And smart remove is called on `a`. In this case `cb` should be removed but not `b->ca->d->3rd_party_file`

Chocobo1 replied:
>And smart remove is called on a. In this case cb should be removed but not b->ca->d->3rd_party_file

OK for me if that is intended.

@glassez 
Just wanted to notify you because now it seems a bit different to what you said at: https://github.com/qbittorrent/qBittorrent/pull/11660#issuecomment-565709115

glassez replied:
I meant the files. Deleting an empty subfolder probably shouldn't be a problem. But I do not have any personal preferences in this problem, because I prefer not to interfere with automated processes.

sledgehammer999 commented:
Guys, can you approve again?
